tensorflow/contrib/go/tensor.go

```diff
@@ -0,0 +1,763 @@
+	cBytesUint16    = 2
+)
+
+// DataType Type of the data contained by a Tensor
+type DataType pb.DataType
+
+// TensorInt Interface to be implemented by the tensors.
+type TensorInt interface {
+	Data() []byte
+	DataSize() int64
+	DataType() DataType
+	GetVal(d ...int) (val interface{}, err error)
+
+	Dim(n int) int
+	NumDims() int
+
+	AsBool() (res []bool, err error)
+	AsFloat32() (res []float32, err error)
+	AsFloat64() (res []float64, err error)
+	AsInt32() (res []int32, err error)
+	AsInt64() (res []int64, err error)
+	AsStr() (res [][]byte, err error)
+
+	String() string
+}
+
+// Tensor Holds a multi-dimensional array of elements of a single data type.
+type Tensor struct {
+	pb.TensorProto
+
+	tensor      TF_Tensor
+	dimWeights  []int
+	memReleased bool
+}
+
+// TensorShape represents the shapre of a Tensor.
+type TensorShape [][]int64
+
+var (
+	// DtInvalid Invalid tensor DataType.
+	DtInvalid = DataType(0)
+	// DtBfloat corresponds to TF_BFLOAT16.
+	DtBfloat = DataType(TF_BFLOAT16)
+	// DtBool corresponds to TF_BOOL.
+	DtBool = DataType(TF_BOOL)
+	// DtComplex corresponds to TF_COMPLEX.
+	DtComplex = DataType(TF_COMPLEX)
+	// DtFloat corresponds to TF_FLOAT.
+	DtFloat = DataType(TF_FLOAT)
+	// DtDouble corresponds to TF_DOUBLE.
+	DtDouble = DataType(TF_DOUBLE)
+	// DtInt8 corresponds to TF_INT8.
+	DtInt8 = DataType(TF_INT8)
+	// DtInt16 corresponds to TF_INT16.
+	DtInt16 = DataType(TF_INT16)
+	// DtInt32 corresponds to TF_INT32.
+	DtInt32 = DataType(TF_INT32)
+	// DtInt64 corresponds to TF_INT64.
+	DtInt64 = DataType(TF_INT64)
+	// DtQint16 corresponds to TF_QINT16.
+	DtQint16 = DataType(TF_QINT16)
+	// DtQuint16 corresponds to TF_QUINT16.
+	DtQuint16 = DataType(TF_QUINT16)
+	// DtQuint32 corresponds to TF_QINT32.
+	DtQuint32 = DataType(TF_QINT32)
+	// DtQint8 corresponds to TF_QINT8.
+	DtQint8 = DataType(TF_QINT8)
+	// DtQuint8 corresponds to TF_QUINT8.
+	DtQuint8 = DataType(TF_QUINT8)
+	// DtString corresponds to TF_STRING.
+	DtString = DataType(TF_STRING)
+	// DtUint8 corresponds to TF_UINT8.
+	DtUint8 = DataType(TF_UINT8)
+	// DtUint16 corresponds to TF_UINT16.
+	DtUint16 = DataType(TF_UINT16)
+)
+
+// NewTensorWithShape returns a new tensor with teh specified type, shape and data.
+// The supported  data types are:
+//  - DtInt8
+//  - DtInt16
+//  - DtInt32
+//  - DtInt64
+//  - DtUint8
+//  - DtUint16
+//  - DtFloat
+//  - DtDouble
+func NewTensorWithShape(shape TensorShape, data interface{}) (*Tensor, error) {
+	v := reflect.ValueOf(data)
+	if v.Kind() != reflect.Slice {
+		return nil, &ErrSliceExpected{
+			dataType: v.Kind().String(),
+		}
+	}
+
+	dataType, err := getDataTypeFromReflect(v.Type().Elem().Kind(), int64(v.Type().Elem().Size()))
+	if err != nil {
+		return nil, err
+	}
+
+	dataSize := int64(v.Len()) * int64(v.Type().Elem().Size())
+	dataPtr := v.Pointer()
+
+	return newTensor(dataType, shape, unsafe.Pointer(dataPtr), dataSize)
+}
+
+// NewTensor Initializes a tensor based on the slice passed by parameter, the
```

Comment:
Comment modified and added some examples inside the comment, I also added code examples for the slice and scalar type on example_test.go:

```go
// NewTensor creates a new Tensor that contains the specified data. The data type
// and shape is deduced from the data parameter.
// ex:
//  - NewTensor("hello") // Creates scalar Tensor of type DTString
//  - NewTensor([]int32{1, 2, 3}) // Creates a 1-D Tensor of type DTInt32
//  - NewTensor([][]float32{{1, 2}, {3, 4}}) // Creates a 2-D Tensor of type DTFloat
```